klippa/src/serialize.rs

```diff
@@ -0,0 +1,378 @@
+        }
+
+        //TODO: add support for clear?
+        let ret = self.head;
+        self.head += size;
+        Ok(ret)
+    }
+
+    #[inline]
+    pub(crate) fn successful(&self) -> bool {
+        !self.errors
+    }
+
+    #[inline]
+    pub(crate) fn in_error(&self) -> bool {
+        !!self.errors
+    }
+
+    #[inline]
+    pub(crate) fn err(&mut self, error_type: SerializeErrorFlags) -> SerializeErrorFlags {
+        self.errors |= error_type;
+        self.errors
+    }
+
+    pub(crate) fn push(&mut self) -> Result<(), SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        let obj_idx = self.object_pool.alloc();
+        let Some(obj) = self.object_pool.get_obj(obj_idx) else {
+            return Err(self.err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER));
+        };
+
+        obj.head = self.head;
+        obj.tail = self.tail;
+        obj.next_obj = self.current;
+        self.current = obj_idx;
+
+        Ok(())
+    }
+
+    pub(crate) fn copy_bytes(&self) -> Vec<u8> {
+        if !self.successful() {
+            return Vec::new();
+        }
+        let len = (self.head - self.start) + (self.end - self.tail);
+        if len == 0 {
+            return Vec::new();
+        }
+
+        let mut out = Vec::with_capacity(len);
+        out.extend_from_slice(self.data.get(self.start..self.head).unwrap());
+        out.extend_from_slice(self.data.get(self.tail..self.end).unwrap());
+        out
+    }
+}
+
+#[allow(dead_code)]
+struct ObjectWrap {
+    pub obj: Object,
+    pub next: usize,
+}
+
+impl Default for ObjectWrap {
+    fn default() -> Self {
+        Self {
+            obj: Object::default(),
+            next: NULL_IDX,
+        }
+    }
+}
+#[allow(dead_code)]
+struct ObjectPool {
+    chunks: Vec<ObjectWrap>,
+    next: usize,
+}
+
+impl Default for ObjectPool {
+    fn default() -> ObjectPool {
+        ObjectPool::empty()
+    }
+}
+
+#[allow(dead_code)]
+impl ObjectPool {
+    const ALLOC_CHUNKS_LEN: usize = 64;
+    pub fn empty() -> ObjectPool {
+        ObjectPool {
+            chunks: vec![],
+            next: NULL_IDX,
+        }
+    }
+    pub fn alloc(&mut self) -> usize {
+        let len = self.chunks.len();
+        if self.next == NULL_IDX {
+            let new_len = len + Self::ALLOC_CHUNKS_LEN;
+            self.chunks.resize_with(new_len, Default::default);
+
+            for (idx, obj) in self
+                .chunks
+                .get_mut(len..new_len - 1)
+                .unwrap()
+                .iter_mut()
+                .enumerate()
+            {
+                obj.next = len + idx + 1;
+            }
+
+            self.chunks.last_mut().unwrap().next = NULL_IDX;
+            self.next = len;
+        }
+
+        let ret = self.next;
+        self.next = self.chunks[self.next].next;
+
+        ret
+    }
+
+    pub fn release(&mut self, obj: usize) {
+        if obj == NULL_IDX || obj >= self.chunks.len() {
```

Comment:
If you use self.chunks.get_mut(obj) then you can handle this case using the returned Option. That avoids having to do the bounds check twice (here and later in self.chunks[obj])